NGS2_WITNESS_Cycle1_experiment2_prep.R

```diff
@@ -0,0 +1,257 @@
+                         sep = '\t', stringsAsFactors = FALSE)
+    bb_ids$bbid <- do.call(c, lapply(strsplit(bb_ids$ROUTER_URL, '/'), function(x) {
+        return(x[length(x)])
+    }))
+    names(bb_ids)[grep('^EMPLOYEE', names(bb_ids))] <- 'empanel_id'
+    return(bb_ids[, c('empanel_id', 'bbid')])
+}
+
+read_input_files <- function(dir) {
+    # identifies all csv files in a directory, reads then into r, and appends them
+    res <- list()
+    files <- list.files(dir, pattern = '.csv$')
+    sources <- strsplit(files, split = '_')
+    for(i in 1:length(files)) {
+        tmp <- read.csv(paste(dir, files[i], sep = '/'), header = TRUE,
+                        sep = ',', stringsAsFactors = FALSE)
+        tmp$source <- paste(sources[[i]][1:2], collapse = '_')
+        res[[i]] <- tmp
+        names(res)[i] <- paste(sources[[i]][1:2], collapse = '_')
+    }
+    return(res)
+}
+
+# Load raw breadboard data
+exp2 <- read_input_files(paste0('NGS2-Cycle', CYCLE, '-Experiment2/data'))
+
+# gather empanelment/breadboard crosswalk information
+empanel_bb_xwalk <- build_empanelment_bb_xwalk('oms_url_upload_20170821_1145.txt')
+
+# Experiment 2
+###############################################################################
+# Restructure raw data output from Breadboard into 'cooperation' dataset
+###############################################################################
+# Subset cooperation decisions, drop "datetime" and "event fields"
+coop <- lapply(exp2, function(x) {
+    return(subset(x, event == 'CooperationDecision',
+                  select = c('id', 'data.name', 'data.value')))
+})
+
+# Reshape long to wide on cooperation events
+coop.events <- lapply(coop, function(x) {
+    tmp <- reshape(x, timevar = "data.name", idvar = c("id"), direction = "wide")
+
+    # Declare `data.value` fields as numerics so events can be sorted logically
+    tmp[, grep('oup|ion|und$', names(tmp))] <- apply(
+        tmp[, grep('oup|ion|und$', names(tmp))], 2, function(y) {
+            return(ifelse(grepl('^_', y), as.numeric(gsub('_', '', y)),
+                          as.numeric(y)))
+        }
+    )
+    return(tmp)
+})
+
+# Create list of group selections by each individual ('pid') at the "ChooseGroup"
+# and "ChangeGroup" events
+group <- lapply(exp2, function(x) {
+    return(subset(x, event %in% c('ChooseGroup', 'ChangeGroup') &
+           data.name %in% c('group', 'pid', 'curRound'),
+           select = c('id', 'data.name', 'data.value')))
+})
+group.events <- lapply(group, function(x) {
+    tmp <- reshape(x, timevar = "data.name", idvar = c("id"),
+                   direction = "wide")[, -1]
+    tmp$link <- (as.numeric(as.factor(tmp$data.value.pid)) * 1000 +
+                 as.numeric(tmp$data.value.curRound))
+    tmp <- tmp[, grep('pid|group|link$', names(tmp))]
+    names(tmp) <- gsub('data.value.', '', names(tmp))
+    return(tmp)
+})
+
+# Since group changes only happen at rounds 1, 4, 8, 12 and so on, need to
```

Comment:
Alright; this is definitely something we'll need to get clarified. This is code I left basically untouched... seems like changes can happen at random during rounds? If so, this block of code will need a definite change, since these breaks are hard-coded.

Reply:
Mark confirmed that individuals are offered a group change round at random, with a 25% chance every round, independently for every individual.